FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/EkmLoginViewModel.kt

```diff
@@ -0,0 +1,160 @@
+/*
+ * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
+ * Contributors:
+ *  DenBond7
+ *  Ivan Pizhenko
+ */
+
+package com.flowcrypt.email.jetpack.viewmodel
+
+import android.app.Application
+import android.content.Context
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.viewModelScope
+import com.flowcrypt.email.R
+import com.flowcrypt.email.api.retrofit.ApiName
+import com.flowcrypt.email.api.retrofit.ApiRepository
+import com.flowcrypt.email.api.retrofit.FlowcryptApiRepository
+import com.flowcrypt.email.api.retrofit.request.api.DomainRulesRequest
+import com.flowcrypt.email.api.retrofit.request.api.LoginRequest
+import com.flowcrypt.email.api.retrofit.request.model.LoginModel
+import com.flowcrypt.email.api.retrofit.response.api.LoginResponse
+import com.flowcrypt.email.api.retrofit.response.base.ApiError
+import com.flowcrypt.email.api.retrofit.response.base.ApiResponse
+import com.flowcrypt.email.api.retrofit.response.base.Result
+import com.flowcrypt.email.api.retrofit.response.model.OrgRules
+import com.flowcrypt.email.util.exception.EkmNotSupportedException
+import com.flowcrypt.email.util.exception.OrgRulesCombinationNotSupportedException
+import kotlinx.coroutines.launch
+
+/**
+ * This [androidx.lifecycle.ViewModel] can be used to resolve domain rules for enterprise users.
+ *
+ * @author Denis Bondarenko
+ *         Date: 10/23/19
+ *         Time: 12:36 PM
+ *         E-mail: DenBond7@gmail.com
+ */
+class EkmLoginViewModel(application: Application) : BaseAndroidViewModel(application) {
+  private val repository: ApiRepository = FlowcryptApiRepository()
+  val ekmLiveData: MutableLiveData<Result<ApiResponse>?> = MutableLiveData()
+
+  fun fetchPrvKeys(account: String, uuid: String, tokenId: String) {
+    ekmLiveData.value = Result.loading()
+    val context: Context = getApplication()
+
+    viewModelScope.launch {
+      ekmLiveData.value = Result.loading(progressMsg = context.getString(R.string.loading))
+      try {
+        val loginResult = repository.login(
+          context,
+          LoginRequest(ApiName.POST_LOGIN, LoginModel(account, uuid), tokenId)
+        )
+
+        when (loginResult.status) {
+          Result.Status.ERROR, Result.Status.EXCEPTION -> {
+            ekmLiveData.value = loginResult
+            return@launch
+          }
+
+          Result.Status.SUCCESS -> {
```

Comment:
I wish these were written with async / await, that would be a lot more readable :-) but I presume it's not so easy here?

Reply:
We use coroutines here. But I use the common logic of how we do it over the whole app. `repository.login` makes a network call with coroutines and returns a parsed results(here we also handle different errors)